jvm-agent/src/main/org/jetbrains/lincheck/jvm/agent/analysis/controlflow/BasicBlockControlFlowGraph.kt

```diff
@@ -97,20 +114,100 @@ class BasicBlockControlFlowGraph(
         val node = instructions.get(idx)
         return (node as? LabelNode)?.label
     }
-    
+
     /**
      * Computes loop-related information for this method.
      */
     fun computeLoopInformation(): MethodLoopsInformation {
         if (loopInfo == null) {
-            // TODO: this is a stub: loop detection is not implemented yet; the method returns empty sites.
-            loopInfo = MethodLoopsInformation().apply {
-                loops.forEach { it.validateLoopEdgesInvariants() }
+            // compute predecessors for all basic blocks
+            val allPredecessors: Array<MutableSet<BasicBlockIndex>> = Array(basicBlocks.size) { mutableSetOf() }
+            val normalPredecessors: Array<MutableSet<BasicBlockIndex>> = Array(basicBlocks.size) { mutableSetOf() }
+            for (e in edges) {
+                val u = e.source
+                val v = e.target
+                allPredecessors[v].add(u)
+                if (e.label !is EdgeLabel.Exception) {
+                    normalPredecessors[v].add(u)
+                }
+            }
+            // compute loop information
+            val dominators = computeDominators(allPredecessors)
```

Comment:
Same observation about dominators and back-edges: they are re-computed multiple times in different methods.

We can add `dominators` and `backEdges` properties directly to CFG, similarly how it is done for `loopInfo`. They should be initially `null`, and they should be initialized after a call to `computeLoopInformation()`.